File: DemoObjC/RDDetailedActionControllerDemoObjC/ViewController.m
//
//  ViewController.m
//  RDDetailedActionControllerDemoObjC
//
//  Created by Firstiar Noorwinanto on 7/23/18.
//  Copyright © 2018 Radical Dreamers. All rights reserved.
//

#import "ViewController.h"
#import <RDDetailedActionController/RDDetailedActionController-Swift.h>

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 4;
}

- (NSString*)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return @"Pick one of these examples";
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 36;
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section {
    return 1;
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ItemCell"];
    
    switch (indexPath.row) {
        case 0:
            cell.textLabel.text = @"Title Only";
            cell.detailTextLabel.text = @"Show an actionsheet with title only item";
            break;
        case 1:
            cell.textLabel.text = @"Icon and Title";
            cell.detailTextLabel.text = @"Show an actionsheet with icon and title";
            break;
        case 2:
            cell.textLabel.text = @"Icon, Title, and Subtitle";
            cell.detailTextLabel.text = @"Show an actionsheet with icon, title, and subtitle";
            break;
        case 3:
            cell.textLabel.text = @"With custom title view";
            cell.detailTextLabel.text = @"Show an actionsheet with custom title view";
            break;
        default:
            break;
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    RDDetailedActionController.defaultTitleFont = [UIFont fontWithName:@"HelveticaNeue" size:14];
    RDDetailedActionController.defaultTitleColor = [UIColor blueColor];
    RDDetailedActionController.defaultTitleSeparatorWidth = 3;
    
    RDDetailedActionView.defaultTitleColor = [UIColor darkGrayColor];
    RDDetailedActionView.defaultSubtitleColor = [UIColor grayColor];

    __block RDDetailedActionController *detailedActionController = nil;
    UIView *titleView = nil;
    UILabel *titleLabel = nil;
    
    switch (indexPath.row) {
        case 0:
            detailedActionController = [[RDDetailedActionController alloc] initWithTitle:@"Select Action"
                                                                                subtitle:@"for selected item"];
            
            [detailedActionController addActionWithTitle:@"Item #1" subtitle:nil icon:nil action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #1 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #2" subtitle:nil icon:nil action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #2 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #3" subtitle:nil icon:nil action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #3 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #4" subtitle:nil icon:nil titleColor:[UIColor redColor] subtitleColor:nil action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #4 clicked");
            }];
            break;
        case 1:
            detailedActionController = [[RDDetailedActionController alloc] initWithTitle:@"Select Action"
                                                                                subtitle:@"for selected item"];
            
            [detailedActionController addActionWithTitle:@"Item #1" subtitle:nil icon:[UIImage imageNamed:@"Image-1"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #1 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #2" subtitle:nil icon:[UIImage imageNamed:@"Image-2"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #2 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #3" subtitle:nil icon:[UIImage imageNamed:@"Image-3"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #3 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #4" subtitle:nil icon:[UIImage imageNamed:@"Image-4"] titleColor:[UIColor redColor] subtitleColor:nil action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #4 clicked");
            }];
            break;
        case 2:
            detailedActionController = [[RDDetailedActionController alloc] initWithTitle:@"Select Action"
                                                                                subtitle:@"for selected item"];
            
            [detailedActionController addActionWithTitle:@"Item #1" subtitle:@"A simple action for that item" icon:[UIImage imageNamed:@"Image-1"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #1 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #2" subtitle:@"A more detailed action for that item" icon:[UIImage imageNamed:@"Image-2"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #2 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #3" subtitle:@"A detailed action with extra steps" icon:[UIImage imageNamed:@"Image-3"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #3 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #4" subtitle:@"A super detailed action with complete extra steps" icon:[UIImage imageNamed:@"Image-4"] titleColor:[UIColor redColor] subtitleColor:[UIColor colorWithRed:1 green:0.4 blue:0.4 alpha:1] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #4 clicked");
            }];
            break;
        case 3:{
            titleView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, 180)];
            titleView.backgroundColor = [UIColor lightGrayColor];

            titleLabel = [[UILabel alloc] initWithFrame:CGRectMake(12, 10, titleView.bounds.size.width - 24, 50)];
            titleLabel.font = [UIFont boldSystemFontOfSize:20];
            titleLabel.numberOfLines = 2;
            titleLabel.textColor = [UIColor whiteColor];
            titleLabel.backgroundColor = [UIColor grayColor];
            titleLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleBottomMargin;
            titleLabel.text = @"This is an action controller utilizing custom title view.";
            [titleView addSubview:titleLabel];
            
            detailedActionController = [[RDDetailedActionController alloc] initWithTitleView:titleView sidePadding:[NSNumber numberWithInt:12]];
            
            [detailedActionController addActionWithTitle:@"Item #1" subtitle:@"A simple action for that item" icon:[UIImage imageNamed:@"Image-1"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #1 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #2" subtitle:@"A more detailed action for that item" icon:[UIImage imageNamed:@"Image-2"] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #2 clicked");
            }];
            
            RDDetailedActionView *child1 = [[RDDetailedActionView alloc] initWithTitle:@"Sub Item #1" subtitle:@"Sub A detailed action with extra steps 1" icon:[UIImage imageNamed:@"Image-3"]  disabled:true action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #3 clicked");
            }];
            
            RDDetailedActionView *child2 = [[RDDetailedActionView alloc] initWithTitle:@"Sub Item #2" subtitle:@"Sub A detailed action with extra steps 2" icon:[UIImage imageNamed:@"Image-3"]  disabled:false action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #3 clicked");
            }];
            NSMutableArray *acts = [NSMutableArray arrayWithObjects:child1, child2, nil];
            [detailedActionController addActionWithTitle:@"Item #3" subtitle:@"A detailed action with extra steps" icon:[UIImage imageNamed:@"Image-3"] children:acts action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #3 clicked");
            }];
            [detailedActionController addActionWithTitle:@"Item #4" subtitle:@"A super detailed action with complete extra steps" icon:[UIImage imageNamed:@"Image-4"] titleColor:[UIColor redColor] subtitleColor:[UIColor colorWithRed:1 green:0.4 blue:0.4 alpha:1] action:^(RDDetailedActionView *actionView) {
                NSLog(@"Item #4 clicked");
            }];
            break;
        }
        default:
            break;
    }
    
    [detailedActionController show];
}

@end

File: DemoSwift/RDDetailedActionControllerDemoSwift/ViewController.swift
//
//  ViewController.swift
//  RDDetailedActionDemo
//
//  Created by Firstiar Noorwinanto on 7/23/18.
//  Copyright © 2018 Radical Dreamers. All rights reserved.
//

import UIKit
import RDDetailedActionController

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 4
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "Pick one of these examples"
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 36
    }
    
    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: "ItemCell")
        
        switch indexPath.row {
        case 0:
            cell?.textLabel?.text = "Title Only"
            cell?.detailTextLabel?.text = "Show an actionsheet with title only item"
            break
        case 1:
            cell?.textLabel?.text = "Icon and Title"
            cell?.detailTextLabel?.text = "Show an actionsheet with icon and title"
            break
        case 2:
            cell?.textLabel?.text = "Icon, Title, and Subtitle"
            cell?.detailTextLabel?.text = "Show an actionsheet with icon, title, and subtitle"
            break
        case 3:
            cell?.textLabel?.text = "Title View With Title Only actions"
            cell?.detailTextLabel?.text = "Show an actionsheet with title view and title only"
            break
        default:
            break
        }
        
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        RDDetailedActionController.defaultTitleFont = UIFont(name: "HelveticaNeue", size: 14)!
        RDDetailedActionController.defaultTitleColor = .blue
        
        RDDetailedActionView.defaultTitleColor = .darkGray
        RDDetailedActionView.defaultSubtitleColor = .gray

        if indexPath.row != 3 {
            let detailedActionController = RDDetailedActionController(title: "Select Action", subtitle: "for selected item")

            switch indexPath.row {
            case 0:
                detailedActionController.addAction(title: "Item #1", subtitle: nil, icon: nil, action: { (actionView) in
                    print("Item #1 clicked")
                })
                detailedActionController.addAction(title: "Item #2", subtitle: nil, icon: nil, action: { (actionView) in
                    print("Item #2 clicked")
                })
                detailedActionController.addAction(title: "Item #3", subtitle: nil, icon: nil, action: { (actionView) in
                    print("Item #3 clicked")
                })
                detailedActionController.addAction(title: "Item #4", subtitle: nil, icon: nil, titleColor: .red, subtitleColor: nil, action: { (actionView) in
                    print("Item #4 clicked")
                })
                break
            case 1:
                detailedActionController.addAction(title: "Item #1", subtitle: nil, icon: UIImage(named: "Image-1"), action: { (actionView) in
                    print("Item #1 clicked")
                })
                detailedActionController.addAction(title: "Item #2", subtitle: nil, icon: UIImage(named: "Image-2"), action: { (actionView) in
                    print("Item #2 clicked")
                })
                detailedActionController.addAction(title: "Item #3", subtitle: nil, icon: UIImage(named: "Image-3"), action: { (actionView) in
                    print("Item #3 clicked")
                })
                detailedActionController.addAction(title: "Item #4", subtitle: nil, icon: UIImage(named: "Image-4"), titleColor: .red, subtitleColor: nil, action: { (actionView) in
                    print("Item #4 clicked")
                })
                break
            case 2:
                detailedActionController.addAction(title: "Item #1", subtitle: "A simple action for that item", icon: UIImage(named: "Image-1"), action: { (actionView) in
                    print("Item #1 clicked")
                })
                detailedActionController.addAction(title: "Item #2", subtitle: "A more detailed action for that item", icon: UIImage(named: "Image-2"), action: { (actionView) in
                    print("Item #2 clicked")
                })
                detailedActionController.addAction(title: "Item #3", subtitle: "A detailed action with extra steps", icon: UIImage(named: "Image-3"), action: { (actionView) in
                    print("Item #3 clicked")
                })
                detailedActionController.addAction(title: "Item #4", subtitle: "A destructor button to remove that item", icon: UIImage(named: "Image-4"), titleColor: .red, subtitleColor: UIColor.init(red: 1, green: 0.4, blue: 0.4, alpha: 1), action: { (actionView) in
                    print("Item #4 clicked")
                })
                break
            default:
                break
            }

            detailedActionController.show()
        }
        else {
            let titleView = UIView(frame: CGRect(x: 0, y: 0, width: self.view.bounds.width, height: 180))
            titleView.backgroundColor = .lightGray
            
            let titleLabel = UILabel(frame: CGRect(x: 12, y: 10, width: titleView.bounds.width - 24, height: 50))
            titleLabel.font = UIFont.boldSystemFont(ofSize: 20)
            titleLabel.numberOfLines = 2
            titleLabel.textColor = .white
            titleLabel.backgroundColor = .gray
            titleLabel.autoresizingMask = [.flexibleWidth, .flexibleBottomMargin]
            titleLabel.text = "This is an action controller utilizing custom title view."
            titleView.addSubview(titleLabel)
            
            let detailedActionController = RDDetailedActionController(titleView: titleView, sidePadding: 12)
            
            detailedActionController.addAction(title: "Item #1", subtitle: nil, icon: nil, action: { (actionView) in
                print("Item #1 clicked")
            })
            detailedActionController.addAction(title: "Item #2", subtitle: nil, icon: nil, action: { (actionView) in
                print("Item #2 clicked")
            })
            detailedActionController.addAction(title: "Item #3", subtitle: nil, icon: nil, action: { (actionView) in
                print("Item #3 clicked")
            })
            detailedActionController.addAction(title: "Item #4", subtitle: nil, icon: nil, titleColor: .red, subtitleColor: nil, action: { (actionView) in
                print("Item #4 clicked")
            })
            
            detailedActionController.show()
        }
    }
}

